src/network/networkd-dhcp-server.c

```diff
@@ -570,3 +579,106 @@ int config_parse_dhcp_server_lpr_servers(
                                                    &n->dhcp_server_lpr, &n->n_dhcp_server_lpr);
 
 }
+
+int config_parse_dhcp_static_leases(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        _cleanup_(sd_dhcp_static_lease_unrefp) sd_dhcp_static_lease *lease = NULL;
+        _cleanup_free_ char *word = NULL;
+        _cleanup_free_ struct ether_addr *n = NULL;
+        OrderedHashmap **static_leases = data;
+        union in_addr_union addr;
+        const char *p;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (isempty(rvalue)) {
+                *static_leases = ordered_hashmap_free(*static_leases);
+                return 0;
+        }
+
+        p = rvalue;
+        r = extract_first_word(&p, &word, " ", 0);
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Invalid hardware address, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        n = new(struct ether_addr, 1);
+        if (!n)
+                return log_oom();
+
+        r = ether_addr_from_string(word, n);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Not a valid MAC address, ignoring: %s", word);
+                return 0;
+        }
+
+        word = mfree(word);
+        r = extract_first_word(&p, &word, " ", 0);
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Invalid IP address, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        r = in_addr_from_string(AF_INET, word, &addr);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
+                           "Failed to parse DHCPv4 IPv4 address data, ignoring assignment: %s", word);
+                return 0;
+        }
+
+        DHCPClientId *c = new(DHCPClientId, 1);
+        if (!c)
+                return log_oom();
+        c->data = malloc0(ETH_ALEN + 1);
+        if (!c->data) {
+                free(c);
+                return log_oom();
+        }
+        ((uint8_t*)  c->data)[0] = 0x01;
+        memcpy((uint8_t*) c->data + 1, n->ether_addr_octet, ETH_ALEN);
+        c->length = ETH_ALEN + 1;
+
+        r = sd_dhcp_static_lease_new(c, &addr.in, &lease);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r,
```

Comment:
if this error path is hit, then `c` is leaked. I figure `c` should be decalred with some form of `_cleanup_` macro use, to ensure it's automatically freeed when the block is left.